blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+
+![Normalization points](./normalization_points.svg)
+
+It seems there are three points where we could intervene and do the de-normalization instead:
+
+1. Retrieve the data ready for display:
+  
+   This can only be done by changing from a simple REST based API to an API that provides access to all data with one call. GraphQL would be an answer, but most often this API is not provided.
+
+2. Put the data de-normalized into the store:
+  
+   This obviously leads to data duplication and problems when updating the data. It's simply not what we want.
+
+3. Select the data ready-for-display from the store:
+
+   This way the data is normalized in the store and updates can be done in an efficient way but there have to be special selectors that simplify selecting all accumulated data. - Let's build one!
+
+## Using Data-Views for De-Normalization
+
+Considering our current example, what we really want to select from the store is the following entity:
+
+```ts
+export interface BookView {
+  title: string;
+  description: string;
+  authors: Author[];
+  tags: Tag[];
+}
+```
+
+It would contain all data that is available on the book entity as well as de-normalized data for authors and tags which we can easily access and iterate over on the template.
+
+To achieve this we can build a new selector that implements this composition by re-using all previous mentioned selectors:
+
+```ts
+export const getBookView = createSelector(
+  getBook,
+  getAuthorsOfBook,
+  getTagsOfBook,
+  (book, authors, tags): BookView =>
+    book && {
+      title: book.title,
+      description: book.description,
+      authors,
+      tags,
+    }
+)
+```
+
+And it will be used in the component and template like this:
+
+```ts
+book$ = this.store.pipe(select(getBookView, this.bookId));
+```
+
+```html
+<div *ngIf="book$ | async as book">
+  <div>{{ book.title }}</div>
+  <div><span *ngFor="let a of book.authors">{{ a.lastName }}</span></div>
+  <div>{{ book.description }}</div>
+  <div><span *ngFor="let t of book.tags">{{ t.name }}</span></div>
+</div>
+```
+
+This way of implementing it provides a good balance of advantages and disadvantages. We push the de-normalization further back so that it doesn't have to be handled in components and templates every time, That way code duplication is prevented. Also, the state still effectively mirrors the API and this simplifies managing and updating entities. The normalization in the data layer also makes it quite easy to later compose selectors for specific tasks, like:
+
+- search for books of a specific author
+- find books with a specific tag
+
+The composition of selectors can however affect memoization, because composed selectors fire every time their input changes. If those inputs are not correctly memoized, the resulting selector will also not be properly memoized.
```

Comment:
+ "We will dig deeper into this in the next section."

Otherwise it ends too abruptly and I immediately was like "Okay, you told us there was a problem and now you just feed me with a Stackblitz instead of an explanation?!"

Reply:
Good point